File: examples/js/example_src/complex.js
/**
 * Complex JavaScript module with more sophisticated functions.
 */

/**
 * Custom error for data validation
 */
class DataValidationError extends Error {
  constructor(message) {
    super(message);
    this.name = 'DataValidationError';
  }
}

/**
 * Calculate the average of an array of numbers
 * @param {number[]} numbers - Array of numbers
 * @returns {number} Average value
 * @throws {Error} If the array is empty
 */
function calculateAverage(numbers) {
  if (!numbers || numbers.length === 0) {
    throw new Error("Cannot calculate average of empty array");
  }
  
  const sum = numbers.reduce((acc, val) => acc + val, 0);
  return sum / numbers.length;
}

/**
 * Validate an email address using regex
 * @param {string} email - Email address to validate
 * @returns {boolean} True if valid, false otherwise
 */
function validateEmail(email) {
  const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  return pattern.test(email);
}

/**
 * Process user data with validation and transformation
 * @param {Object} data - User data object
 * @param {string} data.name - User's name
 * @param {string} data.email - User's email
 * @param {number} data.age - User's age
 * @param {string[]} [data.interests] - User's interests (optional)
 * @returns {Object} Processed data
 * @throws {DataValidationError} If required fields are missing or invalid
 */
function processUserData(data) {
  // Validate required fields
  const requiredFields = ['name', 'email', 'age'];
  for (const field of requiredFields) {
    if (!(field in data)) {
      throw new DataValidationError(`Missing required field: ${field}`);
    }
  }
  
  // Validate email
  if (!validateEmail(data.email)) {
    throw new DataValidationError(`Invalid email: ${data.email}`);
  }
  
  // Validate age
  if (typeof data.age !== 'number' || data.age < 0) {
    throw new DataValidationError(`Invalid age: ${data.age}`);
  }
  
  // Process data
  const result = {
    name: data.name.trim().split(' ').map(part => 
      part.charAt(0).toUpperCase() + part.slice(1).toLowerCase()
    ).join(' '),
    email: data.email.toLowerCase(),
    age: data.age,
    isAdult: data.age >= 18,
    processedAt: new Date().toISOString(),
  };
  
  // Add optional fields
  if (data.interests && Array.isArray(data.interests)) {
    result.interests = data.interests.map(interest => interest.toLowerCase());
  }
  
  // Generate user ID
  result.userId = hashString(`${result.name}:${result.email}`);
  
  console.log(`Processed user data for ${result.name} (ID: ${result.userId})`);
  return result;
}

/**
 * Simple string hashing function (for demo purposes)
 * @param {string} str - String to hash
 * @returns {string} Hashed string
 */
function hashString(str) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash; // Convert to 32bit integer
  }
  return Math.abs(hash).toString(16);
}

/**
 * Analyze the sentiment of a text (mock implementation)
 * @param {string} text - Text to analyze
 * @returns {[string, number]} Tuple of [sentiment, confidence]
 */
function analyzeTextSentiment(text) {
  const positiveWords = ['good', 'great', 'excellent', 'awesome', 'happy', 'love', 'best'];
  const negativeWords = ['bad', 'terrible', 'awful', 'sad', 'hate', 'worst', 'poor'];
  
  const textLower = text.toLowerCase();
  const words = textLower.match(/\w+/g) || [];
  
  const positiveCount = words.filter(word => positiveWords.includes(word)).length;
  const negativeCount = words.filter(word => negativeWords.includes(word)).length;
  
  const totalSentimentWords = positiveCount + negativeCount;
  if (totalSentimentWords === 0) {
    return ['neutral', 0.5];
  }
  
  if (positiveCount > negativeCount) {
    const confidence = positiveCount / totalSentimentWords;
    return ['positive', confidence];
  } else if (negativeCount > positiveCount) {
    const confidence = negativeCount / totalSentimentWords;
    return ['negative', confidence];
  } else {
    return ['neutral', 0.5];
  }
}

/**
 * Create a memoized version of a function with expiration
 * @param {Function} fn - Function to memoize
 * @param {number} [expirySeconds=60] - Time in seconds before cache expires
 * @returns {Function} Memoized function
 */
function memoize(fn, expirySeconds = 60) {
  const cache = new Map();
  
  return function(...args) {
    const key = JSON.stringify(args);
    
    if (cache.has(key)) {
      const [result, timestamp] = cache.get(key);
      if ((Date.now() - timestamp) / 1000 < expirySeconds) {
        console.log(`Cache hit for ${fn.name}`);
        return result;
      }
    }
    
    const result = fn.apply(this, args);
    cache.set(key, [result, Date.now()]);
    console.log(`Cache miss for ${fn.name}, storing result`);
    return result;
  };
}

/**
 * Perform an expensive computation (simulated)
 * @param {number} n - Input parameter
 * @returns {number} Result of computation
 */
function expensiveComputation(n) {
  console.log(`Performing expensive computation with n=${n}`);
  
  // Simulate expensive computation
  let result = 0;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      result += i * j;
    }
  }
  
  return result;
}

// Create memoized version of the expensive computation
const memoizedComputation = memoize(expensiveComputation, 300);

module.exports = {
  DataValidationError,
  calculateAverage,
  validateEmail,
  processUserData,
  analyzeTextSentiment,
  memoize,
  expensiveComputation,
  memoizedComputation
};

File: examples/go/example_src/complex.go
package complex

import (
	"errors"
	"fmt"
	"regexp"
	"strings"
	"time"
)

// DataValidationError represents an error that occurs during data validation.
type DataValidationError struct {
	Field   string
	Message string
}

// Error returns the string representation of the error.
func (e *DataValidationError) Error() string {
	return fmt.Sprintf("Validation error for field '%s': %s", e.Field, e.Message)
}

// NewDataValidationError creates a new DataValidationError.
func NewDataValidationError(field, message string) *DataValidationError {
	return &DataValidationError{
		Field:   field,
		Message: message,
	}
}

// CalculateAverage calculates the average of a slice of numbers.
// Returns an error if the slice is empty.
func CalculateAverage(numbers []float64) (float64, error) {
	if len(numbers) == 0 {
		return 0, errors.New("cannot calculate average of empty slice")
	}

	sum := 0.0
	for _, num := range numbers {
		sum += num
	}

	return sum / float64(len(numbers)), nil
}

// ValidateEmail validates an email address using regex.
func ValidateEmail(email string) bool {
	pattern := `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
	regex := regexp.MustCompile(pattern)
	return regex.MatchString(email)
}

// UserData represents user information.
type UserData struct {
	Name      string   `json:"name"`
	Email     string   `json:"email"`
	Age       int      `json:"age"`
	IsAdult   bool     `json:"is_adult,omitempty"`
	Interests []string `json:"interests,omitempty"`
	UserID    string   `json:"user_id,omitempty"`
	ProcessedAt string `json:"processed_at,omitempty"`
}

// ProcessUserData processes user data with validation and transformation.
func ProcessUserData(data UserData) (UserData, error) {
	// Validate required fields
	if data.Name == "" {
		return UserData{}, NewDataValidationError("name", "Name is required")
	}
	
	if data.Email == "" {
		return UserData{}, NewDataValidationError("email", "Email is required")
	}
	
	if !ValidateEmail(data.Email) {
		return UserData{}, NewDataValidationError("email", fmt.Sprintf("Invalid email: %s", data.Email))
	}
	
	if data.Age < 0 {
		return UserData{}, NewDataValidationError("age", fmt.Sprintf("Invalid age: %d", data.Age))
	}
	
	// Process data
	result := UserData{
		Name:        strings.TrimSpace(data.Name),
		Email:       strings.ToLower(data.Email),
		Age:         data.Age,
		IsAdult:     data.Age >= 18,
		ProcessedAt: time.Now().Format(time.RFC3339),
	}
	
	// Capitalize first letter of each word in name
	nameParts := strings.Fields(result.Name)
	for i, part := range nameParts {
		if len(part) > 0 {
			nameParts[i] = strings.ToUpper(part[:1]) + strings.ToLower(part[1:])
		}
	}
	result.Name = strings.Join(nameParts, " ")
	
	// Add optional fields
	if len(data.Interests) > 0 {
		result.Interests = make([]string, len(data.Interests))
		for i, interest := range data.Interests {
			result.Interests[i] = strings.ToLower(interest)
		}
	}
	
	// Generate user ID (simplified for example)
	result.UserID = fmt.Sprintf("%x", strings.NewReader(result.Name+":"+result.Email).Size())
	
	fmt.Printf("Processed user data for %s (ID: %s)\n", result.Name, result.UserID)
	return result, nil
}

// SentimentAnalysis represents the result of sentiment analysis.
type SentimentAnalysis struct {
	Sentiment  string  `json:"sentiment"`
	Confidence float64 `json:"confidence"`
}

// AnalyzeTextSentiment analyzes the sentiment of a text (mock implementation).
func AnalyzeTextSentiment(text string) SentimentAnalysis {
	positiveWords := []string{"good", "great", "excellent", "awesome", "happy", "love", "best"}
	negativeWords := []string{"bad", "terrible", "awful", "sad", "hate", "worst", "poor"}
	
	textLower := strings.ToLower(text)
	words := regexp.MustCompile(`\w+`).FindAllString(textLower, -1)
	
	positiveCount := 0
	negativeCount := 0
	
	for _, word := range words {
		for _, positive := range positiveWords {
			if word == positive {
				positiveCount++
				break
			}
		}
		
		for _, negative := range negativeWords {
			if word == negative {
				negativeCount++
				break
			}
		}
	}
	
	totalSentimentWords := positiveCount + negativeCount
	if totalSentimentWords == 0 {
		return SentimentAnalysis{
			Sentiment:  "neutral",
			Confidence: 0.5,
		}
	}
	
	if positiveCount > negativeCount {
		return SentimentAnalysis{
			Sentiment:  "positive",
			Confidence: float64(positiveCount) / float64(totalSentimentWords),
		}
	} else if negativeCount > positiveCount {
		return SentimentAnalysis{
			Sentiment:  "negative",
			Confidence: float64(negativeCount) / float64(totalSentimentWords),
		}
	} else {
		return SentimentAnalysis{
			Sentiment:  "neutral",
			Confidence: 0.5,
		}
	}
}

// CacheItem represents an item in the cache.
type CacheItem struct {
	Value     interface{}
	ExpiresAt time.Time
}

// Cache represents a simple in-memory cache with expiration.
type Cache struct {
	data map[string]CacheItem
}

// NewCache creates a new cache.
func NewCache() *Cache {
	return &Cache{
		data: make(map[string]CacheItem),
	}
}

// Set stores a value in the cache with an expiration time.
func (c *Cache) Set(key string, value interface{}, expirySeconds int) {
	c.data[key] = CacheItem{
		Value:     value,
		ExpiresAt: time.Now().Add(time.Duration(expirySeconds) * time.Second),
	}
}

// Get retrieves a value from the cache.
func (c *Cache) Get(key string) (interface{}, bool) {
	item, exists := c.data[key]
	if !exists {
		return nil, false
	}
	
	if time.Now().After(item.ExpiresAt) {
		delete(c.data, key)
		return nil, false
	}
	
	return item.Value, true
}

// ExpensiveComputation performs an expensive computation (simulated).
func ExpensiveComputation(n int) int {
	fmt.Printf("Performing expensive computation with n=%d\n", n)
	
	// Simulate expensive computation
	result := 0
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			result += i * j
		}
	}
	
	return result
}